### ISSUE TITLE
TF400813: Resource not available for anonymous access when switching to https

### ISSUE THREAD

**Issue by geepy** [Microsoft/azure-pipelines-agent#2214]
I am using TFS 2017 Update 3 on premise with separate machines for DB, application tier and build system.
I have successfully registered multiple build agents (v.next) that all run as a windows service. They all run fine.
Today I installed the TFS 3.1 update and the accompanying patch. With this I enabled https access to the application server and changed the public address to https.
No I am trying to switch the build agents to https, but that does not work.
On http connection (http://myserver:8080/tfs as configured url for the agent) everything works fine.
When I try to register the agent for the new connection (https://myserver:8443/tfs), the agent log shows messages like this:
`
[2019-04-12 12:35:39Z ERR  JobDispatcher] Microsoft.VisualStudio.Services.Common.VssUnauthorizedException: TF400813: Resource not available for anonymous access. Client authentication required.
`

What do I have to configure in TFS to make https conections work for build agents?

**Comment by TingluoHuang** [comment]
@geepy  can you share some log when you re-config to Https that contains the failure?

**Comment by TingluoHuang** [comment]
see the following link helps or not. :)
https://docs.microsoft.com/en-us/azure/devops/pipelines/troubleshooting?view=azure-devops#config-fails-while-testing-agent-connection-on-premises-tfs-only

**Comment by geepy** [comment]
> @geepy can you share some log when you re-config to Https that contains the failure?

Here is a worker log which cannot access server resource after succesfully receiving the job details (so in my opinion it must have talked successfully to the server before)
[anonymized_worker.log](https://github.com/Microsoft/azure-pipelines-agent/files/3078635/anonymized_worker.log)

**Comment by geepy** [comment]
> see the following link helps or not. :)
> https://docs.microsoft.com/en-us/azure/devops/pipelines/troubleshooting?view=azure-devops#config-fails-while-testing-agent-connection-on-premises-tfs-only

Hi, unfortunately that didn't help because the registration of the agent works fine. It is shown as "online" in the TFS build queue summary page. It just fails when getting some resource from the server

**Comment by geepy** [comment]
I now received the event logs from my server ond can provide more detail.
When I register my service (or restart it, that has the same effect), builds work for some time. After that, I find the entry
`[2019-04-15 07:57:43Z WARN VisualStudioServices] Authentication failed with status code 401.
Date: Mon, 15 Apr 2019 07:57:42 GMT
P3P: CP="CAO DSP COR ADMa DEV CONo TELo CUR PSA PSD TAI IVDo OUR SAMi BUS DEM NAV STA UNI COM INT PHY ONL FIN PUR LOC CNT"
Server: Microsoft-IIS/8.5
WWW-Authenticate: Bearer, Basic realm="https://_myserver_:8443/tfs", Negotiate, NTLM
X-TFS-ProcessId: 873e3eb7-2fcf-4e3d-979d-019cd1441792
ActivityId: d1ab69a8-2a33-4be0-bc36-da2af6974cce
X-TFS-Session: a854f239-11cd-4c35-b800-ce8ba6d5dfdd
X-VSS-E2EID: 7bf0eaa1-cad5-4eec-a15e-0cbd115df0e3
X-FRAME-OPTIONS: SAMEORIGIN
X-TFS-SoapException: %3c%3fxml+version%3d%221.0%22+encoding%3d%22utf-8%22%3f%3e%3csoap%3aEnvelope+xmlns%3asoap%3d%22http%3a%2f%2fwww.w3.org%2f2003%2f05%2fsoap-envelope%22%3e%3csoap%3aBody%3e%3csoap%3aFault%3e%3csoap%3aCode%3e%3csoap%3aValue%3esoap%3aReceiver%3c%2fsoap%3aValue%3e%3csoap%3aSubcode%3e%3csoap%3aValue%3eUnauthorizedRequestException%3c%2fsoap%3aValue%3e%3c%2fsoap%3aSubcode%3e%3c%2fsoap%3aCode%3e%3csoap%3aReason%3e%3csoap%3aText+xml%3alang%3d%22en%22%3eTF400813%3a+Resource+not+available+for+anonymous+access.+Client+authentication+required.%3c%2fsoap%3aText%3e%3c%2fsoap%3aReason%3e%3c%2fsoap%3aFault%3e%3c%2fsoap%3aBody%3e%3c%2fsoap%3aEnvelope%3e
X-TFS-ServiceError: TF400813%3a+Resource+not+available+for+anonymous+access.+Client+authentication+required.
X-Powered-By: ASP.NET
Lfs-Authenticate: NTLM
X-Content-Type-Options: nosniff
`
in my _diag\worker*.txt.
On the server in TFS debug log I find the following error:
`Microsoft.VisualStudio.Services.WebApi.Jwt.InvalidAudienceException: The audience of the token is invalid. at Microsoft.VisualStudio.Services.WebApi.Jwt.JsonWebTokenUtilities.ValidateAudience(JsonWebToken token, JsonWebTokenValidationParameters parameters) at Microsoft.VisualStudio.Services.WebApi.Jwt.JsonWebTokenUtilities.ValidateToken(JsonWebToken token, JsonWebTokenValidationParameters parameters) at Microsoft.VisualStudio.Services.DelegatedAuthorization.PlatformDelegatedAuthorizationService.IsValidClientCredentialsToken(IVssRequestContext requestContext, Registration registration, DelegatedAuthorizationSettings settings, JsonWebToken jwt)`
So it seems to be related to some authentication token that loses it's validity. But which, and how, and how can I fix this?

Again: one build always runs fine after restarting the build agent service, and I have not found yet a rule after which event (period of time, others) it fails...

**Comment by TingluoHuang** [comment]
@geepy how many TFS AT instances do you have for your server? ex, your TFS configuration includes a load balancer which is backed by 3 TFS AT (application tier) instances.
if that is the case, make sure you update the public url setting on all your TFS AT instances.

if that's not the case, you might need to use Fiddler to capture the network traffic from the agent/worker process and decode the JWT auth token from the http request header.

**Comment by geepy** [comment]
Hi,

I did some research and came up with two approvements:
I logged in with the services' account.
1. I deleted the directory %APPDATA%\local\Microsoft\Team Foundation and it's subdirectories
2. I opened Visual Studio and deleted all (194!) workspaces the team account had created so far
3. I reregistered all agents to the https url

After a gazillion repetitions of this task I came up with a (at leat preliminary) runnung system.

The problem arose because oud TFS instance had not been configured with https before, and I turned that on during the TFS service pack installation. There I switched the public URL, to https too (we do have two App Tier servers, but no load balancer; it seemde to me that changing the public url affects both machins simultaneously).

Now the cache folder contained both urls (the http and https url) to the very same server, which drove the build clients crazy.

When I had that fixed, I intermittently got a "error in XML at position (1,1)" error at the start of the build.
This comes from the workspaces, as these are duplicated, too (one for http and another for https, for the same physical location on the build server). And when you call "tf.exe workspaces", as the build agent does, tf.exe  complains about two workspace mappings on the same server, which is only a warning, but is prepended to the xml delivered to stdout, and that drove the tf.exe output parser mad.

At the momen my build agents can successfully process builds.I'll monitor this for the next 2 days and will surely come back and post my results.


`
`
